plugins/org.jboss.tools.openshift.io.ui/src/main/java/org/jboss/tools/openshift/io/internal/ui/DefaultLoginProvider.java

```diff
@@ -25,23 +31,80 @@
  */
 public class DefaultLoginProvider implements LoginProvider {
 
+	private class LoginJob extends UIJob {
+		private boolean runninginUI = false;
+		private boolean shouldRun = true;
+		private ICluster cluster;
+		private IAccount account;
+		private LoginResponse response;
+		
+		public LoginJob(ICluster cluster, IAccount account) {
+			super("Logging to OpenShift.io");
+			this.cluster = cluster;
+			this.account = account;
+		}
+
+		@Override
+		public IStatus runInUIThread(IProgressMonitor monitor) {
+			if (shouldRun && !monitor.isCanceled()) {
+				runninginUI = true;
+				response = loginInUI(cluster, account);
+				return new Status(IStatus.OK, OpenShiftIOUIActivator.PLUGIN_ID, null);
+			}
+			return new Status(IStatus.CANCEL, OpenShiftIOUIActivator.PLUGIN_ID, null);
+		}
+
+		public boolean isRunninginUI() {
+			return runninginUI;
+		}
+		
+		public void setShouldRun(boolean shouldRun) {
+			this.shouldRun = shouldRun;
+		}
+
+		public LoginResponse getResponse() {
+			return response;
+		}
+	}
+	
 	@Override
 	public LoginResponse login(ICluster cluster, IAccount account) {
 		if (null == Display.getCurrent()) {
-			final LoginResponse response[] = new LoginResponse[1];
-			Display.getDefault().syncExec(new Runnable() {
-				@Override
-				public void run() {
-					response[0] = loginInUI(cluster, account);
-				}
-			});
-			return response[0];
-			
+			return runInJob(cluster, account);
 		} else {
 			return loginInUI(cluster, account);
 		}
 	}
 
+	LoginResponse runInJob(ICluster cluster, IAccount account) {
+		LoginJob job = new LoginJob(cluster, account);
+		job.schedule();
+		IProgressMonitor monitor = new NullProgressMonitor();
+		try {
+			if (job.join(10000, monitor)) {
+				return job.getResponse();
+			} else {
+				throw new InterruptedException();
```

Comment:
what else than us here is throwing InterruptedException (Eclipse shutting down?)? If it's just us then I'm wondering if we couldn't simply have the code of l92 directly following here without the throw.
Furthermore I see us trying to join twice before we finally give up:
1) l84
2) l94
What is the reason for trying twice?

Reply:
The contract of the join method may throw InterruptedException

Reply:
We are trying twice because we want to prevent deadlock (that you may have when an LSP is started when you open the editor). So we first allocate 10s for the job to be on the UI or completed and if we see the job is running on the UI then we wait for the job to finish.